### PR TITLE
Reassemble Space Packets spanning TM frames using the First Header Pointer

### PR DESCRIPTION
**Pull request by lestarch-autobot**
| | |
|:---|:---|
|**_Related Issue(s)_**| Companion to nasa/fprime#5631 |
|**_Has Unit Tests (y/n)_**| y |
|**_Documentation Included (y/n)_**| y (docstrings) |

---
## Change Description

Teaches `SpaceDataLinkFramerDeframer` TM deframing to reassemble Space Packets that span TM transfer frames, using the First Header Pointer (CCSDS 132.0-B-3 4.1.2.7.6):

- On each valid frame, the data field and FHP are passed to a new `reassemble()` step:
  - Continuation bytes before the first header complete the pending spanned packet.
  - `FHP = 0x7FF` (no packet start): the whole field is appended to the pending continuation buffer (bounded by a maximum packet size, discarded with a warning if exceeded).
  - `FHP = 0x7FE` (idle data only): the field is dropped.
  - A trailing partial packet (found by walking Space Packet length tokens via `split_complete_packets()`) is retained for the next frame; only whole packets are emitted downstream.
- Frame loss is detected via virtual channel frame count discontinuity: pending continuation data is discarded and deframing resynchronizes at the frame's FHP.
- Continuation bytes ahead of the first header are only used to complete a pending packet; with nothing pending (mid-stream attach, after loss or an oversize discard) they are dropped. The completed packet must split into whole Space Packets, an idle-only frame invalidates pending data, and an FHP past the data field discards the frame.

Note: the deframer now interprets Space Packet lengths within the TM data field, so `test_deframe_valid_frame` was updated to use a realistic Space Packet payload (previously an arbitrary byte pattern); the assertion is unchanged.

## Rationale

nasa/fprime#5631 adds packet spanning to `Svc::ComAggregator`. Previously the chained deframer discarded inner-layer leftovers between reads, so any packet crossing a TM frame boundary was silently lost on the ground. This change makes the GDS reassemble spanned packets and stay synchronized after frame loss. Behavior for non-spanning streams is unchanged (FHP=0, no pending data).

## Testing/Review Recommendations

- New unit tests: two/three-frame spans, frame-loss resync, orphan continuation (mid-stream attach, post-oversize discard), idle-only frame, VC count wrap, multiple packets per frame, FHP inconsistent with pending length, FHP beyond data field. Full suite: 362/362 pass.
- End-to-end: all 11 `test/int` Ref integration tests pass against a Ref build with spanning enabled (81/103 captured frames carried mid-frame FHP values).
- Review focus: FHP edge values, `split_complete_packets()` length walk, and pending-buffer bounds.

## Future Work

None planned; YAMCS-based ground systems reassemble spanned packets natively via `UdpTmFrameLink`.

### CODE REVIEW

File: src/fprime_gds/common/communication/ccsds/space_data_link.py [hunk 1]
```diff
@@ -22,6 +22,15 @@ class SpaceDataLinkFramerDeframer(FramerDeframer):
...
     TC_TRAILER_SIZE = 2
+    SPACE_PACKET_HEADER_SIZE = 6
+
+    # First Header Pointer special values per CCSDS 132.0-B-3 4.1.2.7.6
+    FHP_MASK = 0x7FF
+    FHP_NO_PACKET_START = 0x7FF  # No packet starts in this frame (continuation data only)
+    FHP_IDLE_DATA_ONLY = 0x7FE  # Frame contains only idle data
+
+    # Maximum reassembled packet size: Space Packet header plus maximum length field value plus 1
+    MAX_PACKET_SIZE = 6 + 65536
```
Comment (lestarch-autobot) [inline src/fprime_gds/common/communication/ccsds/space_data_link.py]: [Maintainability] **could fix** `maint-code-duplication`: the literal `6` restates `SPACE_PACKET_HEADER_SIZE` defined eight lines above.

The comment says "Space Packet header plus ..." but the code encodes the header size a second time, so the two can drift apart and a reader must confirm by hand that `6` is the header. Referencing the named constant makes the bound self-describing.

```suggestion
    MAX_PACKET_SIZE = SPACE_PACKET_HEADER_SIZE + 65536
```

File: src/fprime_gds/common/communication/ccsds/space_data_link.py [hunk 2]
```diff
@@ -173,6 +187,66 @@ def deframe(self, data, no_copy=False):
...
+
+        Args:
+            field: TM frame data field bytes
+            first_header_pointer: First Header Pointer value from the frame Data Field Status
+            vc_count: virtual channel frame count of this frame
+        Return:
+            bytes ready for Space Packet deframing (may be empty)
+        """
+        # Detect frame loss via virtual channel frame count discontinuity
+        if self.last_vc_count is not None and vc_count != (self.last_vc_count + 1) % self.SEQUENCE_NUMBER_MAXIMUM:
```
Comment (lestarch-autobot) [inline src/fprime_gds/common/communication/ccsds/space_data_link.py]: [Maintainability] **could fix** `maint-unclear-naming`: `SEQUENCE_NUMBER_MAXIMUM` is the TC frame sequence-number modulus (see `get_sequence_number`), but here it bounds the TM virtual channel frame count.

The two fields only coincidentally share an 8-bit width; nothing in the name says it also governs VC-count wrap, so an engineer retuning the TC sequence number would silently change loss detection. A dedicated `VC_FRAME_COUNT_MAXIMUM = 256` next to the other TM constants (or `FHP_*` block) keeps the two protocol fields independent.

File: src/fprime_gds/common/communication/ccsds/space_data_link.py [hunk 3]
```diff
@@ -173,6 +187,66 @@ def deframe(self, data, no_copy=False):
...
+            if self.pending:
+                print(
+                    "[WARNING] TM frame loss detected. Discarding partial packet data.",
+                    file=sys.stderr,
+                )
+                self.pending = b""
+            # Discard leading continuation bytes of the lost packet, resync at the first header
+            if first_header_pointer < self.FHP_IDLE_DATA_ONLY:
```
Comment (lestarch-autobot) [inline src/fprime_gds/common/communication/ccsds/space_data_link.py]: [Maintainability] **suggestion** `maint-clever-code`: `< FHP_IDLE_DATA_ONLY` means "is a real pointer" only because the two special values happen to be the two largest 11-bit values.

A reader has to know `0x7FE < 0x7FF` to see that `FHP_NO_PACKET_START` is also excluded here; the reasoning is not written down. Naming both exclusions makes the guard match the two `==` checks that follow.

```suggestion
            if first_header_pointer not in (self.FHP_IDLE_DATA_ONLY, self.FHP_NO_PACKET_START):
```

File: src/fprime_gds/common/communication/ccsds/space_data_link.py [hunk 4]
```diff
@@ -152,16 +165,17 @@ def deframe(self, data, no_copy=False):
...
+                vc_count = data[3]
+                first_header_pointer = struct.unpack_from(">H", data, 4)[0] & self.FHP_MASK
+                field = bytes(data[self.TM_HEADER_SIZE : crc_offset])
                 # Consume the fixed size frame
                 data = data[self.frame_size :]
-                return deframed, bytes(data), bytes(discarded)
+                deframed = self.reassemble(field, first_header_pointer, vc_count)
+                if deframed:
+                    return deframed, bytes(data), bytes(discarded)
+                # Nothing to emit from this frame (continuation or idle only): keep processing
```
Comment (lestarch-autobot) [inline src/fprime_gds/common/communication/ccsds/space_data_link.py]: [Maintainability] **could fix** `maint-misleading-comment`: the comment lists two reasons for an empty result, but `reassemble` also returns empty when the frame holds only the start of a packet that spans into the next frame (FHP=0, nothing complete yet).

That third case is the one `test_deframe_spanning_frame_loss_resync` exercises on `frame1`; a reader trusting this comment would not expect a valid FHP=0 frame to yield nothing.

```suggestion
                # Nothing complete to emit from this frame (continuation, idle, or partial packet start only): keep processing
```

File: test/fprime_gds/common/communication/ccsds/test_space_data_link.py [hunk 5]
```diff
@@ -88,3 +113,62 @@ def test_deframe_incorrect_crc(framer_deframer):
...
+    assert deframed == span + idle
+    assert discarded == b""
+
+
+def test_deframe_spanning_frame_loss_resync(framer_deframer):
+    """Pending continuation data is discarded and deframing resyncs at the FHP after frame loss."""
+    span = make_space_packet(0x101, DATA_FIELD_SIZE + 500)
+    packet = make_space_packet(0x102, DATA_FIELD_SIZE - 500)
+    frame1 = make_tm_frame(span[:DATA_FIELD_SIZE], 0, vc_count=0)
+    # Frame with vc_count=1 is lost; frame 2 starts with 500 continuation bytes of another lost packet
```
Comment (lestarch-autobot) [inline test/fprime_gds/common/communication/ccsds/test_space_data_link.py]: [Maintainability] **could fix** `maint-misleading-comment`: the comment says "frame 2" but the variable it describes is `frame3` (vc_count=2).

The test deliberately skips `frame2` to model the lost frame, so "frame 2" reads as the lost one rather than the received one.

```suggestion
    # The frame with vc_count=1 is lost; frame3 starts with 500 continuation bytes of another lost packet
```

File: src/fprime_gds/common/communication/ccsds/space_data_link.py [hunk 6]
```diff
@@ -152,16 +165,17 @@ def deframe(self, data, no_copy=False):
...
             transmitted_crc = struct.unpack_from(">H", data, crc_offset)[0]
             if transmitted_crc == SpaceDataLinkFramerDeframer.CCITT_CRC_FUNCTION(data[:crc_offset]):
-                # CRC is valid, so we return the deframed data
-                deframed_data_len = (
-                    self.frame_size
-                    - self.TM_TRAILER_SIZE
-                    - self.TM_HEADER_SIZE
-                )
-                deframed = bytes(data[self.TM_HEADER_SIZE : self.TM_HEADER_SIZE + deframed_data_len])
+                # CRC is valid: extract the data field and reassemble packets spanning frames
```
Comment (lestarch-autobot) [inline src/fprime_gds/common/communication/ccsds/space_data_link.py:180]: [Documentation] **could fix** `deframe` docstring (line 147, `"""Deframe TM frames"""`) no longer describes the method's contract after this change.

`deframe` is the public `FramerDeframer` entry point and the autodoc surface for this plugin; it is now stateful across calls (`pending`/`last_vc_count`), returns the concatenation of all *complete* Space Packets reassembled so far rather than the raw TM data field, and silently buffers continuation fragments. None of that is visible from the one-line docstring, and the base-class docstring in `framing.py` still promises "one packet or None". Anchored here because line 147 is outside the diff; suggested replacement:

```python
    def deframe(self, data, no_copy=False):
        """Deframe TM frames into complete Space Packets

        Validates each fixed-size TM frame (SCID/VCID, CRC) and passes its data field to `reassemble`, which
        buffers packets that span frames using the First Header Pointer. Returns the concatenated complete
        Space Packets available after the frame (possibly several, none for continuation/idle-only frames),
        the unconsumed bytes, and any discarded bytes. Keeps `pending`/`last_vc_count` state across calls.
        """
```

File: src/fprime_gds/common/communication/ccsds/space_data_link.py [hunk 7]
```diff
@@ -173,6 +187,66 @@ def deframe(self, data, no_copy=False):
...
+                self.pending += field
+                if len(self.pending) > self.MAX_PACKET_SIZE:
+                    print(
+                        "[WARNING] Spanned packet exceeds maximum packet size. Discarding partial packet data.",
+                        file=sys.stderr,
+                    )
+                    self.pending = b""
+            return b""
+        # Continuation bytes before the first header complete the pending spanned packet
+        emitted = self.pending + field[:first_header_pointer]
```
Comment (lestarch-autobot) [inline src/fprime_gds/common/communication/ccsds/space_data_link.py]: [Correctness] **must fix** Leading continuation bytes are emitted even when `self.pending` is empty, so the frame-loss resync does not hold.

If the loss is detected on a continuation-only frame (`FHP == 0x7FF`), `pending` is cleared but nothing is trimmed; the next in-sequence frame (`FHP = N > 0`) then emits `field[:N]` — the tail of the lost packet — ahead of the complete packets. The same happens when the GDS attaches mid-stream (first frame seen has `FHP > 0`) and after the max-size discard. Downstream `SpacePacketFramerDeframer` byte-rotates through the orphan bytes and can match a bogus header that consumes into the valid packets that follow (reproduced with a 4-frame span and frame 1 dropped: the complete 36-byte packet in frame 3 was lost, three 1-byte bogus packets came out). With no pending data there is nothing for the leading bytes to complete, so drop them.

```suggestion
        emitted = (self.pending + field[:first_header_pointer]) if self.pending else b""
```

File: test/fprime_gds/common/communication/ccsds/test_space_data_link.py [hunk 8]
```diff
@@ -88,3 +113,62 @@ def test_deframe_incorrect_crc(framer_deframer):
...
+    frame3 = make_tm_frame(b"\xAA" * 500 + packet, 500, vc_count=2)
+
+    deframed, remaining, discarded = framer_deframer.deframe(frame1)
+    assert deframed is None  # Spanning packet start retained
+    deframed, remaining, discarded = framer_deframer.deframe(frame3)
+    assert deframed == packet  # Lost continuation dropped, resynced at the first header
+    assert remaining == b""
```
Comment (lestarch-autobot) [inline test/fprime_gds/common/communication/ccsds/test_space_data_link.py:174]: [Test Quality] **suggestion** Three new behaviors have no test pinning them: the `FHP_IDLE_DATA_ONLY` branch (must not disturb pending data), the `vc_count` 255→0 wrap in the loss check (must not be treated as loss), and the `split_complete_packets()` length walk over more than one whole packet per frame (every current test has at most one complete packet per frame). All three pass on this head; adding them keeps the FHP edge values and the length walk from regressing silently.

```suggestion
    assert remaining == b""


def test_deframe_idle_only_frame_keeps_pending(framer_deframer):
    """An idle-only frame (FHP=0x7FE) emits nothing and leaves pending continuation data intact."""
    span = make_space_packet(0x101, DATA_FIELD_SIZE + 500)
    idle = make_space_packet(0x7FF, DATA_FIELD_SIZE - 500)
    frame1 = make_tm_frame(span[:DATA_FIELD_SIZE], 0, vc_count=0)
    frame_idle = make_tm_frame(
        bytes(DATA_FIELD_SIZE), SpaceDataLinkFramerDeframer.FHP_IDLE_DATA_ONLY, vc_count=1
    )
    frame2 = make_tm_frame(span[DATA_FIELD_SIZE:] + idle, 500, vc_count=2)

    assert framer_deframer.deframe(frame1)[0] is None
    assert framer_deframer.deframe(frame_idle)[0] is None
    assert framer_deframer.deframe(frame2)[0] == span + idle


def test_deframe_spanning_vc_count_wrap(framer_deframer):
    """A vc_count wrap from 255 to 0 is a valid sequence, not frame loss."""
    span = make_space_packet(0x101, DATA_FIELD_SIZE + 500)
    idle = make_space_packet(0x7FF, DATA_FIELD_SIZE - 500)
    frame1 = make_tm_frame(span[:DATA_FIELD_SIZE], 0, vc_count=255)
    frame2 = make_tm_frame(span[DATA_FIELD_SIZE:] + idle, 500, vc_count=0)

    assert framer_deframer.deframe(frame1)[0] is None
    assert framer_deframer.deframe(frame2)[0] == span + idle


def test_deframe_multiple_packets_in_frame(framer_deframer):
    """split_complete_packets walks several whole packets in one frame."""
    packets = [make_space_packet(0x100 + i, 700) for i in range(3)]
    idle = make_space_packet(0x7FF, DATA_FIELD_SIZE - 3 * 700)
    frame = make_tm_frame(b"".join(packets) + idle, 0, vc_count=0)

    deframed, remaining, discarded = framer_deframer.deframe(frame)
    assert deframed == b"".join(packets) + idle
    assert remaining == b""
```

File: src/fprime_gds/common/communication/ccsds/space_data_link.py (hunk 7 above)
Comment (lestarch-autobot) [inline src/fprime_gds/common/communication/ccsds/space_data_link.py]: [Security] **suggestion** — general-vulnerability/input-validation-gap: when `self.pending` is empty (first frame after start, or right after the `MAX_PACKET_SIZE` discard above), `field[:first_header_pointer]` is the tail of a packet whose header was never received, yet it is emitted as header-aligned data. `SpacePacketFramerDeframer.deframe` then parses those bytes as a Space Packet header; a stray length field there makes it swallow or stall on the real packets that follow in the same frame. The frame-loss branch already drops these leading bytes via the FHP; apply the same rule whenever no packet is pending.

```suggestion
        # Continuation bytes before the first header complete the pending spanned packet;
        # with nothing pending they belong to a packet whose start was never received
        emitted = self.pending + field[:first_header_pointer] if self.pending else b""
```

File: src/fprime_gds/common/communication/ccsds/space_data_link.py [hunk 9]
```diff
@@ -173,6 +187,66 @@ def deframe(self, data, no_copy=False):
...
+        is used to resynchronize to the next packet header.
+
+        Args:
+            field: TM frame data field bytes
+            first_header_pointer: First Header Pointer value from the frame Data Field Status
+            vc_count: virtual channel frame count of this frame
+        Return:
+            bytes ready for Space Packet deframing (may be empty)
+        """
+        # Detect frame loss via virtual channel frame count discontinuity
```
Comment (lestarch-autobot) [inline src/fprime_gds/common/communication/ccsds/space_data_link.py:217]: [Security] **suggestion** — general-vulnerability/unchecked-offset: `first_header_pointer` (0..0x7FD, from the frame header) is used as a slice offset without a range check against `len(field)`, which is only 1016 bytes at the default `TmFrameFixedSize` of 1024. A CRC-valid frame carrying an out-of-range pointer makes `field[:first_header_pointer]` append the entire data field to `self.pending` and emit it as a completed packet, while `field[first_header_pointer:]` is empty. Python slicing keeps this from crashing, but the malformed header corrupts the reassembled stream; reject the frame's data field instead.

```suggestion
        if first_header_pointer < self.FHP_IDLE_DATA_ONLY and first_header_pointer >= len(field):
            print(
                "[WARNING] First Header Pointer beyond TM frame data field. Discarding frame data.",
                file=sys.stderr,
            )
            return b""
        # Detect frame loss via virtual channel frame count discontinuity
```

File: src/fprime_gds/common/communication/ccsds/space_data_link.py (hunk 7 above)
Comment (lestarch-autobot) [inline src/fprime_gds/common/communication/ccsds/space_data_link.py]: [Design] **must fix** Orphan continuation bytes are emitted downstream when there is no pending packet start.

When `self.pending` is empty and `0 < FHP < 0x7FE`, `field[:first_header_pointer]` is the tail of a packet whose header was never seen, yet it is emitted to the Space Packet deframer as if it completed a packet. This is the routine case of the GDS attaching to an already-running flight side (`last_vc_count is None`, so the loss branch is skipped), and it also occurs after the `MAX_PACKET_SIZE` discard above. The design already treats orphan continuation as droppable (the loss-resync branch trims at FHP; the `0x7FF` branch ignores the field when nothing is pending), but this path is inconsistent with that: the downstream deframer receives garbage bytes it must rotate through, and a false header sync on them can swallow the real first packet.

Reproduced at this head: a first frame with `FHP=300` and 300 leading filler bytes returns the 300 filler bytes prepended to the packet.

```suggestion
        emitted = self.pending + field[:first_header_pointer] if self.pending else b""
```

File: src/fprime_gds/common/communication/ccsds/space_data_link.py [hunk 10]
```diff
@@ -173,6 +187,66 @@ def deframe(self, data, no_copy=False):
...
+                    file=sys.stderr,
+                )
+                self.pending = b""
+            # Discard leading continuation bytes of the lost packet, resync at the first header
+            if first_header_pointer < self.FHP_IDLE_DATA_ONLY:
+                field = field[first_header_pointer:]
+                first_header_pointer = 0
+        self.last_vc_count = vc_count
+        if first_header_pointer == self.FHP_IDLE_DATA_ONLY:
+            return b""
```
Comment (lestarch-autobot) [inline src/fprime_gds/common/communication/ccsds/space_data_link.py:229]: [Design] **suggestion** An idle-only frame should invalidate a pending spanned packet.

A packet started in frame *n* cannot continue through an idle-only frame (`FHP=0x7FE`, CCSDS 132.0-B-3 4.1.2.7.6): its continuation would have made frame *n+1* a `0x7FF` continuation frame instead. Returning early here keeps the stale `self.pending`, so the next frame with a mid-field FHP glues the stale bytes to unrelated continuation data and emits them as a packet (reproduced at this head: pending 2214 bytes + idle-only frame + `FHP=50` frame emits 4428 bytes instead of the 2164-byte packet). The F Prime `ComAggregator` in nasa/fprime#5631 does not emit idle-only frames mid-packet, so this only matters for other senders or corrupted status fields, hence non-blocking.

```suggestion
        if first_header_pointer == self.FHP_IDLE_DATA_ONLY:
            # A packet cannot continue through an idle-only frame: any pending start is stale
            self.pending = b""
            return b""
```

File: src/fprime_gds/common/communication/ccsds/space_data_link.py (hunk 7 above)
Comment (lestarch-autobot) [reply src/fprime_gds/common/communication/ccsds/space_data_link.py]: [Test Quality] **Concur** — also in scope for test quality: no test exercises `pending` empty with `first_header_pointer > 0` (mid-stream attach, post-loss, or post-`MAX_PACKET_SIZE` discard); `make_tm_frame(b"\xAA" * 500 + packet, 500, vc_count=7)` on a fresh deframer asserting `deframed == packet` fails on this head and should land with the fix.

Comment (lestarch-autobot) [inline src/fprime_gds/common/communication/ccsds/space_data_link.py]: [Operational] **must fix** Orphan continuation bytes are forwarded as packet data when no packet is pending

ops-contract-trace: `reassemble()` promises "bytes ready for Space Packet deframing", but with `self.pending` empty and FHP > 0 the leading `field[:first_header_pointer]` bytes (tail of a packet whose head was never received) are emitted as if header-aligned.
Reached on the first frame after the GDS attaches to a running deployment (`last_vc_count is None` skips the loss check) and on the first FHP > 0 frame after the `MAX_PACKET_SIZE` discard.
Traced consequence (probe at the test's 2222-byte frames, 500 orphan bytes): `SpacePacketFramerDeframer` accepted a false header inside the orphan bytes and delivered a 1030-byte bogus packet, then hit a length it could not satisfy and left the real 1714-byte packet in `remaining`, which `ChainedFramerDeframer.deframe_all()` discards. With telemetry-like orphan bytes it delivered 6 bogus packets and still lost the real one.
So every GDS start against a spanning-enabled downlink can both inject garbage packets and lose the first real packet(s). Remedy: only prepend continuation bytes to an existing pending packet; drop them otherwise.

```suggestion
        emitted = self.pending + field[:first_header_pointer] if self.pending else b""
```

File: src/fprime_gds/common/communication/ccsds/space_data_link.py [hunk 11]
```diff
@@ -173,6 +187,66 @@ def deframe(self, data, no_copy=False):
...
+                        "[WARNING] Spanned packet exceeds maximum packet size. Discarding partial packet data.",
+                        file=sys.stderr,
+                    )
+                    self.pending = b""
+            return b""
+        # Continuation bytes before the first header complete the pending spanned packet
+        emitted = self.pending + field[:first_header_pointer]
+        # Emit only whole packets; a packet starting in this frame may span into the next
+        complete, self.pending = self.split_complete_packets(field[first_header_pointer:])
+        return emitted + complete
```
Comment (lestarch-autobot) [inline src/fprime_gds/common/communication/ccsds/space_data_link.py:262]: [Operational] **suggestion** Reassembled packet is never checked against its own length; 8-bit VC-count aliasing glues unrelated data silently

ops-failure-path: loss detection is `vc_count != last + 1 (mod 256)`, so a gap of exactly k*256 frames (1 in 256 outages longer than 256 frames, about 260 kB at the default 1016-byte data field) or an FSW reboot when the last count was 255 is invisible, and the pending head is joined to the next frame's `field[:FHP]` without checking that the result is exactly whole packets.
Probe (2222-byte frames, gap of 257 frames): no warning; a 2508-byte packet with the pending packet's valid header but 300 payload bytes from an unrelated packet was delivered to the GDS decoders, plus a 1030-byte bogus packet, and the frame's real 1514-byte packet was lost.
Remedy: verify `emitted` splits into whole packets and drop it with a warning otherwise. This also catches an FHP inconsistent with the pending header from any cause (flight-side bug, FHP > data-field length), and bounds the damage to one warning instead of silent wrong telemetry.

```suggestion
        if self.split_complete_packets(emitted)[1]:
            print(
                "[WARNING] First Header Pointer inconsistent with pending packet length. Discarding partial packet data.",
                file=sys.stderr,
            )
            emitted = b""
        return emitted + complete
```

File: src/fprime_gds/common/communication/ccsds/space_data_link.py [hunk 12]
```diff
@@ -173,6 +187,66 @@ def deframe(self, data, no_copy=False):
...
+
+    @classmethod
+    def split_complete_packets(cls, data):
```
Comment (lestarch-autobot) [inline src/fprime_gds/common/communication/ccsds/space_data_link.py:265]: [Operational] **could fix** Reassembly assumes the TM data field carries Space Packets; the standalone `raw-space-data-link` framing plugin now silently misparses other payloads

ops-config-extreme, operational judgment call: `SpaceDataLinkFramerDeframer` is also registered on its own as the `raw-space-data-link` framing plugin (`fprime_gds/plugin/system.py`). Before this change that configuration passed the data field through untouched; now `split_complete_packets()` reads bytes 4-5 of whatever is in the field as a Space Packet length, so a non-SPP payload (e.g. F Prime packets framed directly into TM by a project framer) is partially withheld as `pending` and re-emitted glued to the next frame's output, corrupting every packet boundary with no warning.
No in-repo flight framer produces that layout (`Svc::Ccsds::TmFramer` always fills with an SPP idle packet), so this is deployment-dependent.
Remedy: state in the class docstring / plugin help that the TM data field must contain Space Packets, or bypass `reassemble()` when the deframer is not used inside the composite.

cc @thomas-bc @LeStarch — low-confidence finding, please confirm.

File: src/fprime_gds/common/communication/ccsds/space_data_link.py (hunk 7 above)
Comment (lestarch-autobot) [reply src/fprime_gds/common/communication/ccsds/space_data_link.py]: [Operational] **Concur** — also in scope for operational consequences: reached on every GDS attach to a running spanning-enabled downlink; probe shows a bogus 1030-byte packet delivered and the frame's real packet dropped via the chained deframer's discarded inner `remaining`.

Comment (lestarch-autobot) [reply src/fprime_gds/common/communication/ccsds/space_data_link.py]: [Summary] **Duplicate** — consolidated into https://github.com/nasa/fprime-gds/pull/338#discussion_r3981987939.

Comment (lestarch-autobot) [reply src/fprime_gds/common/communication/ccsds/space_data_link.py]: [Summary] **Duplicate** — consolidated into https://github.com/nasa/fprime-gds/pull/338#discussion_r3981987939.

Comment (lestarch-autobot) [reply src/fprime_gds/common/communication/ccsds/space_data_link.py]: [Summary] **Duplicate** — consolidated into https://github.com/nasa/fprime-gds/pull/338#discussion_r3981987939.

**Review by lestarch-autobot — COMMENTED** [review]
## Automated review summary  (run 1)

### Per-agent results

| Agent | must fix | suggestion | could fix | future work | outstanding | Verdict |
|---|---:|---:|---:|---:|---:|---|
| Security Vulnerabilities | 0 | 2 | 0 | 0 | 2 | Go |
| Supply Chain / Runner Safety | 0 | 0 | 0 | 0 | 0 | Go |
| F Prime C/C++ Design | 0 | 0 | 0 | 0 | 0 | Go |
| Documentation Currency | 0 | 0 | 1 | 0 | 1 | Go |
| Design | 1 | 1 | 0 | 0 | 2 | No-Go |
| Architecture | ERROR | ERROR | ERROR | ERROR | ERROR | ERROR |
| Test Quality | 1 | 1 | 0 | 0 | 2 | No-Go |
| Correctness | 1 | 0 | 0 | 0 | 1 | No-Go |
| Operational | 1 | 1 | 1 | 0 | 3 | No-Go |
| Maintainability | 0 | 1 | 4 | 0 | 5 | Go |
| **CI safety** | — | — | — | — | — | **Go** |
| **Totals** | 1 | 5 | 6 | 0 | 12 | **No-Go** |

Totals count the orphan-continuation finding once (Correctness canonical; Design, Operational, Security duplicates closed and Test Quality concurrence folded in); per-agent rows keep each reviewer's own counts.

Duplicates consolidated this run: 3 (threads closed by the §5h post-pass)

<details>
<summary>Supply-chain surfaces</summary>

| Surface | Outstanding |
|---|---|
| Dependencies | clean |
| Vendored / submodule | clean |
| Build / test infrastructure | clean |
| Workflows / actions / scripts | clean |
| Generator output | clean |
| Prompt-injection | clean |
| Review-system integrity | clean |

</details>

<details>
<summary>Outstanding must-fix items (1)</summary>

**Correctness**
- Leading continuation bytes (`field[:first_header_pointer]`) are emitted as header-aligned data even when `self.pending` is empty, so the frame-loss / mid-stream-attach resync does not hold and orphan fragments can be parsed as packets — also: Design, Operational, Security, Test Quality — https://github.com/nasa/fprime-gds/pull/338#discussion_r3981987939

</details>

### Merge readiness
**Merge readiness: No-Go** — Architecture reviewer failed to run, and 1 consolidated must-fix (Correctness, concurred by Design, Operational, Test Quality) is outstanding.

Review event: `REQUEST_CHANGES` intended; posted as a comment-type review because GitHub rejects `REQUEST_CHANGES` from the PR author's own identity (`lestarch-autobot`), which this run's token shares.

### Agents that did not run on this PR
- None — all ten reviewers were invoked; Architecture failed (reviewer aborted before authenticating; no metadata review or inline comments posted) and is rendered as an ERROR row above.

---

One resync fix and an Architecture re-run, and this packet reassembly is cleared for downlink.

File: src/fprime_gds/common/communication/ccsds/space_data_link.py (hunk 1 above)
Comment (lestarch-autobot) [reply src/fprime_gds/common/communication/ccsds/space_data_link.py]: Fixed in f34885a: `MAX_PACKET_SIZE = SPACE_PACKET_HEADER_SIZE + 65536`.

*Pushed by Devin AI*

File: src/fprime_gds/common/communication/ccsds/space_data_link.py (hunk 2 above)
Comment (lestarch-autobot) [reply src/fprime_gds/common/communication/ccsds/space_data_link.py]: Fixed in f34885a: added `VC_FRAME_COUNT_MAXIMUM = 256` next to the TM constants and used it for the loss check; `SEQUENCE_NUMBER_MAXIMUM` is TC-only again.

*Pushed by Devin AI*

File: src/fprime_gds/common/communication/ccsds/space_data_link.py (hunk 3 above)
Comment (lestarch-autobot) [reply src/fprime_gds/common/communication/ccsds/space_data_link.py]: Resolved in f34885a by removing the trim block: with the orphan-continuation fix (r3981987939) leading bytes are dropped whenever nothing is pending, so the loss branch only needs to clear `pending` and the `< FHP_IDLE_DATA_ONLY` guard is gone.

*Pushed by Devin AI*

File: src/fprime_gds/common/communication/ccsds/space_data_link.py (hunk 4 above)
Comment (lestarch-autobot) [reply src/fprime_gds/common/communication/ccsds/space_data_link.py]: Fixed in f34885a.

*Pushed by Devin AI*

File: test/fprime_gds/common/communication/ccsds/test_space_data_link.py (hunk 5 above)
Comment (lestarch-autobot) [reply test/fprime_gds/common/communication/ccsds/test_space_data_link.py]: Fixed in f34885a.

*Pushed by Devin AI*

File: src/fprime_gds/common/communication/ccsds/space_data_link.py (hunk 6 above)
Comment (lestarch-autobot) [reply src/fprime_gds/common/communication/ccsds/space_data_link.py:180]: Fixed in f34885a: `deframe` docstring now describes the stateful reassembly and the return contract.

*Pushed by Devin AI*

File: src/fprime_gds/common/communication/ccsds/space_data_link.py (hunk 7 above)
Comment (lestarch-autobot) [reply src/fprime_gds/common/communication/ccsds/space_data_link.py]: Fixed in f34885a: `field[:first_header_pointer]` is only used when `self.pending` is non-empty; otherwise the leading bytes are dropped. Added `test_deframe_orphan_continuation_dropped` (mid-stream attach, the suggested `b"\xAA" * 500 + packet` / FHP=500 / vc_count=7 case) and `test_deframe_orphan_continuation_after_oversize_discard`.

*Pushed by Devin AI*

File: test/fprime_gds/common/communication/ccsds/test_space_data_link.py (hunk 8 above)
Comment (lestarch-autobot) [reply test/fprime_gds/common/communication/ccsds/test_space_data_link.py:174]: Added in f34885a: `test_deframe_spanning_vc_count_wrap`, `test_deframe_multiple_packets_in_frame`, and an idle-only test. The idle-only test asserts the behavior from the design finding (r3982002416) instead: an idle-only frame invalidates pending data, so the stale span tail is dropped and only the packet at the FHP is emitted (`test_deframe_idle_only_frame_discards_pending`).

*Pushed by Devin AI*

File: src/fprime_gds/common/communication/ccsds/space_data_link.py (hunk 9 above)
Comment (lestarch-autobot) [reply src/fprime_gds/common/communication/ccsds/space_data_link.py:217]: Fixed in f34885a: `first_header_pointer >= len(field)` discards the frame data with a warning. It also clears `pending`, since the pending packet cannot be completed from a frame whose pointer is unusable. Covered by `test_deframe_fhp_beyond_data_field` (1024-byte frame, FHP=1016).

*Pushed by Devin AI*

File: src/fprime_gds/common/communication/ccsds/space_data_link.py (hunk 10 above)
Comment (lestarch-autobot) [reply src/fprime_gds/common/communication/ccsds/space_data_link.py:229]: Fixed in f34885a: idle-only frames clear `pending`. Covered by `test_deframe_idle_only_frame_discards_pending`.

*Pushed by Devin AI*

File: src/fprime_gds/common/communication/ccsds/space_data_link.py (hunk 11 above)
Comment (lestarch-autobot) [reply src/fprime_gds/common/communication/ccsds/space_data_link.py:262]: Fixed in f34885a: `pending + field[:FHP]` is checked with `split_complete_packets`; a non-empty remainder discards it with a warning. Covered by `test_deframe_fhp_inconsistent_with_pending`.

*Pushed by Devin AI*

File: src/fprime_gds/common/communication/ccsds/space_data_link.py (hunk 12 above)
Comment (lestarch-autobot) [reply src/fprime_gds/common/communication/ccsds/space_data_link.py:265]: Documented in f34885a: the class docstring now states that TM deframing expects Space Packets in the data field (FHP + Space Packet length walk) and that other contents are not supported. Bypassing `reassemble()` when used standalone was not done since `raw-space-data-link` has no signal for the payload type; happy to add a flag if a non-SPP TM payload is a real configuration.

*Pushed by Devin AI*

**Review by thomas-bc — APPROVED** [review]
